zenoh-flow/src/runtime/dataflow/instance/runners/mod.rs

```diff
@@ -235,50 +238,54 @@ impl NodeRunner {
     ///  # Errors
     /// An error variant is returned in case the run returns an error.
-    fn run_stoppable(&self, signal: Signal) -> ZFResult<()> {
-        async fn run(runner: &NodeRunner) -> RunAction {
-            match runner.run().await {
-                Ok(_) => RunAction::Stop,
-                Err(e) => RunAction::RestartRun(Some(e)),
-            }
-        }
-
-        async fn stop(signal: Signal) -> RunAction {
-            signal.wait().await;
-            RunAction::Stop
-        }
-        async_std::task::block_on(async move {
-            loop {
-                let cloned_signal = signal.clone();
-                match stop(cloned_signal).race(run(self)).await {
-                    RunAction::RestartRun(e) => {
-                        log::error!(
-                            "[Node: {}] The run loop exited with {:?}, restarting…",
-                            self.get_id(),
-                            e
-                        );
-                    }
-                    RunAction::Stop => {
-                        log::trace!(
-                            "[Node: {}] Received kill command, killing runner",
-                            self.get_id()
-                        );
-                        self.stop().await;
-                        return Ok(());
-                    }
+    async fn run_stoppable(&self, signal: Arc<Barrier>) -> ZFResult<()> {
+        loop {
+            log::info!("NodeRunner {} starting run!", self.get_id());
+
+            let my_id = self.get_id();
+            let future_stop = async {
+                log::info!("NodeRunner {} waiting for stop trigger", my_id);
+                signal.wait().await;
+                log::info!("NodeRunner {} received stop trigger", my_id);
+                RunAction::Stop
+            };
+
+            let future_run = async {
+                log::info!("NodeRunner {} running", self.get_id());
+                match self.run().await {
+                    Ok(_) => RunAction::Stop,
+                    Err(e) => RunAction::RestartRun(Some(e)),
+                }
+            };
+
+            match future_stop.race(future_run).await {
+                RunAction::RestartRun(e) => {
+                    log::error!(
+                        "[Node: {}] The run loop exited with {:?}, restarting…",
+                        self.get_id(),
+                        e
+                    );
+                }
+                RunAction::Stop => {
+                    log::trace!(
+                        "[Node: {}] Received kill command, killing runner",
+                        self.get_id()
+                    );
+                    self.stop().await;
+                    return Ok(());
                 }
             }
-        })
+        }
     }
 
     /// Starts the node, returning the `RunnerManager` to stop it.
     pub fn start(&self) -> RunnerManager {
-        let signal = Signal::new();
+        let signal = Arc::new(Barrier::new(2));
```

Comment:
We should explain the logic behind this barrier (in particular why the magic number 2)